plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md

```diff
@@ -0,0 +1,557 @@
+- The `restrained` behavior is only effective on pages where you can place official Coder components. If you are not on one of these pages, the fallback auth UI will always be visible until the user authenticates.
+- Fewer options for customizing the styling
+
+### Authenticating via `useCoderAuth`
+
+The `useCoderAuth` hook provides state and functions for updating Coder authentication state within your Backstage deployment. When called, it gives you back a `CoderAuth` object
+
+```tsx
+// This is a simplified version of the type; the real type is set up as a
+// discriminated union to increase type safety and ergonomics further
+type CoderAuth = Readonly<{
+  status: CoderAuthStatus; // Union of strings
+  token: string | undefined;
+  error: unknown;
+
+  isAuthenticated: boolean;
+  registerNewToken: (newToken: string) => void;
+  ejectToken: () => void;
+}>;
+```
+
+#### Pros
+
+- Gives you the finest level of control over all auth concerns
+- Easy to import into any component
+
+#### Cons
+
+- Zero UI logic out of the box; you have to make all components yourself
+- Must be called within a `CoderProvider` component
+
+## Caching API data for UIs
+
+All core logic in the Coder plugin uses [Tanstack Query v4](https://tanstack.com/query/v4). As it is already a dependency for the Coder plugin, it is highly recommended that you also use the library when building out your own components.
+
+The three main hooks from the library that you will likely use with the SDK are:
+
+- `useQuery`
+- `useMutation`
+- `useQueryClient`
+
+At present, the Coder plugin provides a convenience wrapper for connecting `useQuery` to the Coder SDK and to Coder auth state. This is the `useCoderQuery` hook – this is a great option if a component should only care about making queries and doesn't need to interact directly with auth state.
+
+We are also considering making a `useMutation` wrapper called `useCoderMutation`. If you have any thoughts or requests on how it should behave, please open an issue!
+
+## Why Tanstack Query (aka React Query)?
+
+ui.dev has a phenomenal [article](https://ui.dev/why-react-query) and [video](https://www.youtube.com/watch?v=OrliU0e09io) series that breaks things down more exhaustively, but in short, Tanstack Query:
+
+- Simplifies how API data is shared throughout an application
+- Manages race conditions, canceling requests, and revalidating requests
+- Provides a shared API for both queries and mutations
+
+### Problems with fetching via `useEffect`
+
+All functions returned by the Coder SDK maintain stable memory references for the entire lifetime of the SDK. In that sense, every one of these functions can be safely placed inside a `useEffect` dependency array to perform data fetching. In theory, `usEffect` can be used to trigger API calls, while the results (and their relevant loading/error/success states) can be stored via `useState`.
+
+In practice, however, this setup causes a lot of problems:
```

Comment:
I think we may be over-explaining React internals here which is leading to a lengthier document. What do you think?

Reply:
I agree. IMO this guide should just provide solid, out-of-the-box example components and snippets that people can learn from. If we want to provide break-glass examples, docs, references, trade-offs, and recommendations we should do it via a separate doc. We want to make this process look extremely simple and easy to do without the plugin author to make a bunch of decisions.

Most people I assume will copy & paste our examples and modify it for their case and not decide to do their own querying and stuff. If they break glass, I like us guiding them to be a better programmer but think it definitely needs to be separated.